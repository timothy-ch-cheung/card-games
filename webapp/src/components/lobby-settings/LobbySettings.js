import React, {useEffect, useState} from 'react';
import Card from "react-bootstrap/Card";
import NumberPicker from "../number-picker/NumberPicker";
import GameModes from "../../GameModes";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

function LobbySettings(props) {
    const MAX_ROUNDS_MULTIPLIER = 4;
    const [validated, setValidated] = useState(false);

    const isSingleRoundIncrement = () => {
        return GameModes[props.gameMode] ? GameModes[props.gameMode].roundIncrement : undefined;
    }

    const getStep = (game) => {
        if (!game) {
            return;
        }

        let stepType = game.roundIncrement
        switch (stepType) {
            case "ONE":
                return 1;
            case "PLAYER_COUNT":
                return props.numPlayers;
            default:
                return undefined;
        }
    };

    const submitHandler = e => {
        e.preventDefault();
        const form = e.currentTarget;
        const valid = form.checkValidity();

        if (valid === false || !e.target.numRounds.value) {
            e.stopPropagation();
            setValidated(true);
            return;
        }

        setValidated(false);
        props.onSubmit();
    };

    let step = getStep(GameModes[props.gameMode])
    const [rounds, setRounds] = useState(1);
    const [totalRounds, setTotalRounds] = useState(step);

    const calculateTotalRounds = (step, rounds) => {
        return step * rounds;
    }

    const decrease = () => {
        if (GameModes[props.gameMode] && totalRounds > step) {
            setRounds(rounds - 1);
        }
    }

    const increase = () => {
        if (GameModes[props.gameMode] && totalRounds < (step * MAX_ROUNDS_MULTIPLIER)) {
            setRounds(rounds + 1);
        }
    }

    useEffect(() => {
        setTotalRounds(calculateTotalRounds(step, rounds))
    }, [step, rounds])

    const RoundPicker = () => {
        return (
            <div style={{display: "flex"}}>
                <div style={{marginRight: "5px"}}>
                    <Form.Label>{isSingleRoundIncrement() ? "Rounds" : "Stages"}</Form.Label>
                    <NumberPicker value={rounds} onIncrease={increase} onDecrease={decrease}
                                  name={"numRounds"} data-test={'round-number-picker'}/>
                </div>
                <div style={isSingleRoundIncrement() ? {display: "none"} : {display: "visible"}}>
                    <Form.Label>Rounds</Form.Label>
                    <Form.Control required type="text" name="totalRounds" disabled value={totalRounds}
                                  style={{backgroundColor: "#fff", textAlign: "center", width: "80px"}}
                                  data-test="round-number-text"/>
                </div>
            </div>
        );
    }

    return (
        <Card style={{width: "35%", margin: "10px"}}>
            <Card.Header style={{paddingLeft: "10px"}}>Game: {props.gameMode}</Card.Header>
            <Card.Body style={{padding: "5px 10px", height: "400px", overflowY: "auto"}}>
                <Form noValidate validated={validated} onSubmit={submitHandler} data-test={'lobby-settings-form'}>
                    <RoundPicker/>
                </Form>
            </Card.Body>
            <Card.Body style={{borderTop: "1px solid #dfdfdf", padding: "10px"}}>
                <Button variant="info" type="submit" data-test={"start-game-btn"}>Start</Button>
            </Card.Body>
        </Card>
    );
}

export default LobbySettings;
