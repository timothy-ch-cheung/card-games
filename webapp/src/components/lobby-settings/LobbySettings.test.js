import {configure, mount, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from "react";
import LobbySettings from "./LobbySettings";

configure({adapter: new Adapter()});

describe("TEST SUITE LobbySettings: ", () => {
    let wrapper;

    test('Matches LobbySettings snapshot [Match Two]', () => {
        wrapper = shallow(<LobbySettings numPlayers={2} gameModes="Match Two"/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('Matches LobbySettings snapshot [Choice Poker]', () => {
        wrapper = shallow(<LobbySettings numPlayers={2} gameModes="Choice Poker"/>);
        expect(wrapper).toMatchSnapshot();
    });

    describe('Number of rounds - ', () => {
        test('plus/minus buttons change rounds by player count', () => {
            wrapper = mount(<LobbySettings gameMode={"Match Two"} numPlayers={2}/>);
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(2);
            wrapper.find('button[data-test="numRounds-plus-btn"]').simulate('click');
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(4);
            wrapper.find('button[data-test="numRounds-minus-btn"]').simulate('click');
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(2);
        });

        test('plus/minus buttons change rounds by one', () => {
            wrapper = mount(<LobbySettings gameMode={"Choice Poker"} numPlayers={2}/>);
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(1);
            wrapper.find('button[data-test="numRounds-plus-btn"]').simulate('click');
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(2);
            wrapper.find('button[data-test="numRounds-minus-btn"]').simulate('click');
            expect(wrapper.find('input[data-test="round-number-text"]').props().value).toEqual(1);
        });

        test('onSubmit function is set and called on submit event', () => {
            let submit = jest.fn();
            wrapper = mount(<LobbySettings gameMode={"Match Two"} numPlayers={2} onSubmit={submit}/>);
            const formEventMocked = {preventDefault: jest.fn(), target: {numRounds: {value: 4}}};
            wrapper.find('form').simulate('submit', formEventMocked);
            expect(wrapper.props().onSubmit).toEqual(submit);
            expect(submit).toHaveBeenCalledTimes(1);
        });

        test('onSubmit function is not called when rounds invalid', () => {
            let submit = jest.fn();
            wrapper = mount(<LobbySettings gameMode={"Match Two"} numPlayers={2} onSubmit={submit}/>);
            const formEventMocked = {preventDefault: jest.fn(), target: {numRounds: {value: undefined}}};
            wrapper.find('form').simulate('submit', formEventMocked);
            expect(submit).toHaveBeenCalledTimes(0);
        });
    });
});