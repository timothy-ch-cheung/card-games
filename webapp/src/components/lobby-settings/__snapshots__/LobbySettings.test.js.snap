// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST SUITE LobbySettings:  Matches LobbySettings snapshot [Choice Poker] 1`] = `
<Card
  body={false}
  style={
    Object {
      "margin": "10px",
      "width": "35%",
    }
  }
>
  <CardHeader
    style={
      Object {
        "paddingLeft": "10px",
      }
    }
  >
    Game: 
  </CardHeader>
  <CardBody
    style={
      Object {
        "height": "400px",
        "overflowY": "auto",
        "padding": "5px 10px",
      }
    }
  >
    <Form
      data-test="lobby-settings-form"
      inline={false}
      noValidate={true}
      onSubmit={[Function]}
      validated={false}
    >
      <RoundPicker />
    </Form>
  </CardBody>
  <CardBody
    style={
      Object {
        "borderTop": "1px solid #dfdfdf",
        "padding": "10px",
      }
    }
  >
    <Button
      active={false}
      data-test="start-game-btn"
      disabled={false}
      type="submit"
      variant="info"
    >
      Start
    </Button>
  </CardBody>
</Card>
`;

exports[`TEST SUITE LobbySettings:  Matches LobbySettings snapshot [Match Two] 1`] = `
<Card
  body={false}
  style={
    Object {
      "margin": "10px",
      "width": "35%",
    }
  }
>
  <CardHeader
    style={
      Object {
        "paddingLeft": "10px",
      }
    }
  >
    Game: 
  </CardHeader>
  <CardBody
    style={
      Object {
        "height": "400px",
        "overflowY": "auto",
        "padding": "5px 10px",
      }
    }
  >
    <Form
      data-test="lobby-settings-form"
      inline={false}
      noValidate={true}
      onSubmit={[Function]}
      validated={false}
    >
      <RoundPicker />
    </Form>
  </CardBody>
  <CardBody
    style={
      Object {
        "borderTop": "1px solid #dfdfdf",
        "padding": "10px",
      }
    }
  >
    <Button
      active={false}
      data-test="start-game-btn"
      disabled={false}
      type="submit"
      variant="info"
    >
      Start
    </Button>
  </CardBody>
</Card>
`;
